### ISSUE TITLE
Code Action (Wrap with Widget) moves cursor to parent Widget

### ISSUE THREAD

**Issue by shivahnshankar** [emacs-lsp/lsp-dart#118]
**Describe the bug**
If i keep the cursor (point) on a Widget and select "Wrap with Widget..." from the Code Actions menu on the right, a new widget by name "widget" is created but the cursor now moves to the parent widget. So i have to now scroll down, and rename the "widget" to whatever i want.
In VSCode, the same thing happens but the new parent Widget called "widget" is selected and highlighted, and just typing the new name would change the widget.

**To Reproduce**
Before wrap with widget:
<img width="553" alt="Screenshot 2021-08-28 at 11 33 57 AM" src="https://user-images.githubusercontent.com/8016298/131208260-3fe9e3ef-dba3-443a-901a-e81b64d73d84.png">
After wrap with widget:
<img width="519" alt="Screenshot 2021-08-28 at 11 34 18 AM" src="https://user-images.githubusercontent.com/8016298/131208266-626d4816-caba-4231-b0be-dd513b5ef0c3.png">

**Expected behavior**
The cursor(point) is at the beginning of the new widget for easier editing like in VS Code.
<img width="606" alt="Screenshot 2021-08-28 at 11 43 27 AM" src="https://user-images.githubusercontent.com/8016298/131208311-7c659534-0589-4091-9909-806538b5528c.png">

**Version**
Include here the result of: <kbd>M-x</kbd> `lsp-dart-version`
[LSP Dart] 1.19.0 at 2021.08.28 @ Emacs 27.2
[Dart SDK] Dart SDK version: 2.15.0-65.0.dev (dev) (Fri Aug 27 11:53:11 2021 -0700) on "macos_x64"
Doom Emacs latest version.

**Comment by ericdallo** [comment]
Thanks for the report!

I'm not sure what would cause that, I'll try to repro and understand what server is sending back to client

**Comment by ericdallo** [comment]
I confirmed the issue, maybe Dart-Code's maintainer, @DanTup can help us understand if Dart-Code does anything after that code action is applied?

**Comment by DanTup** [comment]
@ericdallo Yep, LSP doesn't natively support what we need here, so to support this we use an _unofficial_ LSP extension created by the Rust team:

https://github.com/rust-analyzer/rust-analyzer/blob/b35559a2460e7f0b2b79a7029db0c5d4e0acdb44/docs/dev/lsp-extensions.md#snippet-textedit
https://github.com/microsoft/language-server-protocol/issues/724

The server support is here:

https://github.com/dart-lang/sdk/commit/44275c6001837550d4a6b7339237ae1c27bdbc19

And the middleware we inject into VS Code's LSP client was added here:

https://github.com/Dart-Code/Dart-Code/pull/3215

This is an unofficial extension to LSP but if you're happy to support that, there's no reason you couldn't do the same (note: it's not certain if it'll become official LSP, or if it does, whether there will be changes from the current proposal). To support it you need to advertise that you support it (via the `experimental` part of your capabilities, then interpret snippets in the code actions responses.

Let me know if anything isn't clear! :-)

**Comment by ericdallo** [comment]
Oh, now I see, quite a hack :)
Thanks, that really makes sense now, I'll check how Dart-Code do that and try to do something similar from lsp-dart side

**Comment by DanTup** [comment]
I wouldn't call it a hack, it might one day be part of LSP proper and possibly very similar to it's implemented there, but it is a bit awkward. I/we decided it was an annoying enough issues that it was worth implementing even if not standard (although mostly I'd prefer to wait for standards). Unfortunately LSP moves quite slowly (which is not entirely a bad thing given a lot of clients/editors implement it) so I don't know if/when it'll be standard.

**Comment by DanTup** [comment]
Oh btw, we put a flag in for it in VS Code even though it defaults to on, so that if there are ever any issues (for ex. it becomes standard but is slightly different, or we find bugs in the snippet handling/generation) it's easy for users to disable (or us to ship a new default to turn it off without an SDK release).  This prevents the client advertising support for it, which prevents the server generating the snippet version.

If you do implement it, I'd recommend also having an easy switch for users to turn it off, just in case :-)

**Comment by ericdallo** [comment]
yes, I totally agree that makes sense for the LSP spec, thanks for your attention and explanation!

**Comment by yyoncho** [comment]
> @ericdallo Yep, LSP doesn't natively support what we need here, so to support this we use an _unofficial_ LSP extension created by the Rust team:

FTR we support that for rust, so maybe it just have to be enabled for dart.

**Comment by DanTup** [comment]
Probably controlled by this `snippetTextEdit` in `experimental` here?:

https://github.com/emacs-lsp/lsp-mode/blob/8f9259af6fc80a609c2c068d0f59c371205aca89/clients/lsp-rust.el#L803

**Comment by ericdallo** [comment]
@DanTup yes! I just tested adding that flag and it worked perfectly, thank you!
@shivahnshankar available on next 1.19.4 release :)

**Comment by DanTup** [comment]
Great, glad to hear! Do let me know if you hit any issues with it.
